### PR TITLE
Fix race

### PR DESCRIPTION
**Pull request by bcolloran**
this fixes the race condition in #971, lays groundwork for iframe reset, removes some import cycles (and sets the stage to remove a bunch more)

### CODE REVIEW

File: src/redux-to-port-message-passer.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    // this.postMessage =
+  }
+  addDispatch(dispatch) {
+    this.reduxManager.addMsgFn(dispatch);
+  }
+
+  postMessage(...params) {
+    this.portManager.sendMsg(...params);
+  }
+  dispatchToRedux(...params) {
```
Comment (hamilton) [inline src/redux-to-port-message-passer.js]: is there a reason we don't just call this `dispatch`?

File: src/redux-to-port-message-passer.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  }
+}
+
+export class MessagePasser {
+  constructor() {
+    this.reduxManager = new MessageQueueManager();
+    this.portManager = new MessageQueueManager();
+  }
+
+  addPostMessage(postMessage) {
```
Comment (hamilton) [inline src/redux-to-port-message-passer.js]: could this be `addPostMessageFunction`?

File: src/redux-to-port-message-passer.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    this.portManager.addMsgFn(postMessage);
+    // this.postMessage =
+  }
+  addDispatch(dispatch) {
```
[the remark this answers is not in the document]
Comment (bcolloran) [reply src/redux-to-port-message-passer.js]: yep, sounds good to me. changed to `connectDispatch` and `connectPostMessage`

File: src/eval-frame/port-to-editor.js [hunk 4]
```diff
@@ -1,35 +1,53 @@
...
+}
+connectToEditor();
 
 export function postMessageToEditor(messageType, message) {
   portToEditor.postMessage({ messageType, message });
 }
 
+messagePasser.addPostMessage(postMessageToEditor);
+
 function receiveMessage(event) {
```
[the remark this answers is not in the document]
Comment (bcolloran) [reply src/eval-frame/port-to-editor.js]: this seems sensible enough. let's file this as a separate code quality issue.

File: src/redux-to-port-message-passer.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      this.msgQueue.push(msg);
+    } else {
+      this.msgQueue = [msg];
+    }
+  }
+}
+
+export class MessagePasser {
+  constructor() {
+    this.reduxManager = new MessageQueueManager();
```
[the remark this answers is not in the document]
Comment (bcolloran) [reply src/redux-to-port-message-passer.js]: i changed `reduxManager` -> `dispatchQueueManager`, and `portManager` -> `postMessageQueueManager` (both are instances of `MessageQueueManager`). How is that?

File: src/redux-to-port-message-passer.js (hunk 2 above)
Comment (bcolloran) [reply src/redux-to-port-message-passer.js]: i changed this one to `connectPostMessage` to match `connectDispatch`. that work?

File: src/redux-to-port-message-passer.js (hunk 1 above)
Comment (bcolloran) [reply src/redux-to-port-message-passer.js]: hmm i guess not really :-) i'll change it

File: src/eval-frame/port-to-editor.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (bcolloran) [reply src/eval-frame/port-to-editor.js]: filed #1450

**Comment by bcolloran** [comment]
Thanks for the review @hamilton! I totally agree that the message passing between iframes needs more work. filed #1450 for that. I also made the other changes you requested -- this is ready for another review pass

File: src/eval-frame/port-to-editor.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (hamilton) [reply src/eval-frame/port-to-editor.js]: great, will resolve this issue

File: src/redux-to-port-message-passer.js (hunk 2 above)
Comment (hamilton) [reply src/redux-to-port-message-passer.js]: great

**Pull request merged**